### ISSUE TITLE
2-Storey Building of Interpreters

### ISSUE THREAD

**Issue by eignnx** [eignnx/rellog#23]
Maybe for load-time macros, the goal is to build an environment that is available to the runtime system.

Should it be functional then? No, maybe just a DCG.

## Import
`import` reads the specified file and outputs a ~module object~ collection of path-qualified relation **definitions**:

Assume `shapes.rellog` is a local file.

```prolog
[import shapes] --> {
    shape::[rect_width][rect_height][rect_area] = ( body )
    shape::[circ_radius][circ_area] = ( body )
}
```

## Documentation
`doc` outputs a `[sig][doc]` fact given the following definition's signature and the provided doc string.

```prolog
[[doc "Relates a list to its length."]]:
[list {}][length 0]

[list {X ..Xs}][Length]
    ( body )


[doc "Relates a list to its length."] --> {
    [sig [list][length]][doc "Relates a list to its length."]
}
```

## Assert/Define
`define` would output a list of definitions that will be available at runtime. Any choicepoints define more clauses in the list?

```prolog
[define [Head][Body]] --> {
    Head = Body
}
```

# Summary
So to define a load-time macro:
1. Write a DCG which outputs clauses
2. Make sure it's available at load-time (TODO: how?)
3. Invoke it using `[[blah]]` meta brackets, possible accepting a clause as argument (TODO: how to accept *clause* as opposed to, like, a *docstring* as argument?)

Regarding point 3 above:

```prolog
[[macro_sig [Doc]][options [NextSig]]]
    - ...
```

Updates #15

**Comment by eignnx** [comment]
```
[macro [import Path]][Options]
    - [[file_path Path][Clauses][Directives]]
    - Clauses
    - Directives
```

(Use double brackets to indicate goal is not part of dcg output. That will require making `[[blah]]` a term not just a top-level item.)

**Comment by eignnx** [comment]
closed by 1f8fc4b
